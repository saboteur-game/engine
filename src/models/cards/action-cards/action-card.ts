import Card from "../card";
import CardParameters from "../card-parameters";
import { Pojo } from "../../../utils";

class ActionCard extends Card {
  parameters: CardParameters | undefined;

  play(parameters: CardParameters): ActionCard {
    super.play(parameters);
    this.parameters = parameters;
    return this;
  }

  toJS(): Pojo {
    return {
      ...super.toJS(),
      parameters: this.parameters,
    };
  }
}

export default ActionCard;
