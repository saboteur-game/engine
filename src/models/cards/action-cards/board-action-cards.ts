import { ActionCard } from ".";
import CardParameters from "../card-parameters";

export class BoardActionCard extends ActionCard {
  play(parameters: CardParameters): BoardActionCard {
    return super.play(parameters);
  }
}

export class MapActionCard extends BoardActionCard {}

export class RockfallActionCard extends BoardActionCard {}
