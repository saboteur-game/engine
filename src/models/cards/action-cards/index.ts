import ActionCard from "./action-card";
import {
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
} from "./board-action-cards";
import {
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
} from "./tool-action-cards";

export {
  ActionCard,
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
};
