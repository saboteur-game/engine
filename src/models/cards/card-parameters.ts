import { Tools } from "../../constants";
import Player from "../player";

export interface IBoardCardParameters {
  position: string;
}

export interface IToolActionCardParameters {
  player: Player;
  appliedTo: Tools;
}

type CardParameters = IToolActionCardParameters | IBoardCardParameters;

export default CardParameters;
