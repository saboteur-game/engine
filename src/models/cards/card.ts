import { generateId, Pojo } from "../../utils";
import CardParameters from "./card-parameters";

export enum Status {
  unused = "unused",
  played = "played",
  discarded = "discarded",
}

class Card {
  id: string;
  status: Status;

  constructor() {
    this.id = generateId();
    this.status = Status.unused;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  play(parameters: CardParameters): Card {
    return this;
  }

  setDiscarded(): void {
    // TODO: Check if the state was unused and throw if it was anything else
    this.status = Status.discarded;
  }

  setPlayed(): void {
    // TODO: Check if the state was unused or played and throw if it was anything else
    this.status = Status.played;
  }

  toJS(): Pojo {
    return {
      id: this.id,
      status: this.status,
    };
  }
}

export default Card;
