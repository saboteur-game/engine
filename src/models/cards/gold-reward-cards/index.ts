import { generateId, Pojo } from "../../../utils";

class GoldRewardCard {
  id: string;
  value: number;

  constructor(value: 1 | 2 | 3) {
    this.id = generateId();
    this.value = value;
  }

  toJS(): Pojo {
    return {
      id: this.id,
      value: this.value,
    };
  }
}

export default GoldRewardCard;
