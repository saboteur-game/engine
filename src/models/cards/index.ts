import { shuffle } from "../../utils";
import { Tools } from "../../constants";
import Card, { Sides } from "./card";
import {
  StartPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
  PassageCard,
  DeadendCard,
  PathCard,
} from "./path-cards";
import {
  MapActionCard,
  RockfallActionCard,
  BreakActionCard,
  RepairActionCard,
  ActionCard,
} from "./action-cards";

const multiply = (length: number, getValue: () => Card) =>
  new Array(length).fill(undefined).map(() => getValue());
const randomBoolean = () => Math.random() < 0.5;

/*
 * Saboteur cards
 * - 44 path cards
 *  -- 1 starting card
 *  -- 3 finishing cards
 *    --- 1 gold [Connected, 1-2-3-4]
 *    --- 2 rock
 *       ---- 1 [Connected 1-2]
 *       ---- 1 [Connected 1-4]
 *  -- 40 tunnel cards
 *    --- 31 passage cards
 *       ---- 3 passage [Connected 2-4]
 *       ---- 4 passage [Connected 1-3]
 *       ---- 4 passage [Connected 1-4]
 *       ---- 5 passage [Connected 1-2]
 *       ---- 5 passage [Connected 1-2-4]
 *       ---- 5 passage [Connected 1-2-3]
 *       ---- 5 passage [Connected 1-2-3-4]
 *    --- 9 dead-end cards
 *       ---- 1 dead-end [Connected 1]
 *       ---- 1 dead-end [Connected 2]
 *       ---- 1 dead-end [Connected 1-2]
 *       ---- 1 dead-end [Connected 1-3]
 *       ---- 1 dead-end [Connected 1-4]
 *       ---- 1 dead-end [Connected 2-4]
 *       ---- 1 dead-end [Connected 1-2-3]
 *       ---- 1 dead-end [Connected 1-2-4]
 *       ---- 1 dead-end [Connected 1-2-3-4]
 * - 27 action cards
 *  -- 6 map cards
 *  -- 3 rock-fall cards
 *  -- 9 break tool cards
 *    --- 3 pick
 *    --- 3 lamp
 *    --- 3 wagon
 *  -- 9 repair tool cards
 *    --- 2 pick
 *    --- 2 lamp
 *    --- 2 wagon
 *    --- 1 pick and lamp
 *    --- 1 pick and wagon
 *    --- 1 lamp and wagon
 *
 * Notes: Card connections
 *
 *        (1)
 *     +-------+
 *     |       |
 * (4) |       | (2)
 *     |       |
 *     +-------+
 *        (3)
 */

export interface IPlacedCards {
  start: StartPathCard;
  gold: GoldFinishPathCard;
  rock1: RockFinishPathCard;
  rock2: RockFinishPathCard;
}

export const getPlacedCards = (): IPlacedCards => {
  const start = new StartPathCard();
  const gold = new GoldFinishPathCard();
  const rock1 = new RockFinishPathCard([Sides.top, Sides.right]);
  const rock2 = new RockFinishPathCard([Sides.top, Sides.left]);

  [start, gold, rock1, rock2].forEach((card) => card.setPlayed());

  return { start, gold, rock1, rock2 };
};

export const getShuffledDeck = (): Array<ActionCard | PathCard> => {
  const pathCards = [
    ...multiply(
      3,
      () => new PassageCard([Sides.right, Sides.left], randomBoolean())
    ),
    ...multiply(
      4,
      () => new PassageCard([Sides.top, Sides.bottom], randomBoolean())
    ),
    ...multiply(
      4,
      () => new PassageCard([Sides.top, Sides.left], randomBoolean())
    ),
    ...multiply(
      5,
      () => new PassageCard([Sides.top, Sides.right], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard([Sides.top, Sides.right, Sides.left], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard([Sides.top, Sides.right, Sides.bottom], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard(
          [Sides.top, Sides.right, Sides.bottom, Sides.left],
          randomBoolean()
        )
    ),
    new DeadendCard([Sides.top], randomBoolean()),
    new DeadendCard([Sides.right], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right], randomBoolean()),
    new DeadendCard([Sides.top, Sides.bottom], randomBoolean()),
    new DeadendCard([Sides.top, Sides.left], randomBoolean()),
    new DeadendCard([Sides.right, Sides.left], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right, Sides.bottom], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right, Sides.left], randomBoolean()),
    new DeadendCard(
      [Sides.top, Sides.right, Sides.bottom, Sides.left],
      randomBoolean()
    ),
  ] as Array<PathCard>;

  const actionCards = [
    ...multiply(6, () => new MapActionCard()),
    ...multiply(3, () => new RockfallActionCard()),
    ...multiply(3, () => new BreakActionCard(Tools.pick)),
    ...multiply(3, () => new BreakActionCard(Tools.lamp)),
    ...multiply(3, () => new BreakActionCard(Tools.wagon)),
    ...multiply(2, () => new RepairActionCard(Tools.pick)),
    ...multiply(2, () => new RepairActionCard(Tools.lamp)),
    ...multiply(2, () => new RepairActionCard(Tools.wagon)),
    new RepairActionCard([Tools.pick, Tools.lamp]),
    new RepairActionCard([Tools.pick, Tools.wagon]),
    new RepairActionCard([Tools.lamp, Tools.wagon]),
  ] as Array<ActionCard>;

  return shuffle(
    ([] as Array<ActionCard | PathCard>).concat(pathCards, actionCards)
  );
};
