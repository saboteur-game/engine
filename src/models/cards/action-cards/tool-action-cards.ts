import { Tools } from "../../../constants";
import { ActionCard } from ".";
import { Pojo } from "../../../utils";
import { IToolActionCardParameters } from "../card-parameters";

export class ToolActionCard extends ActionCard {
  appliesTo: Tools[];

  constructor(appliesTo: Tools | Tools[]) {
    super();
    this.appliesTo = ([] as Tools[]).concat(appliesTo);
  }

  play(parameters: IToolActionCardParameters): ActionCard {
    if (!this.appliesTo.includes(parameters.appliedTo)) {
      throw new Error(
        `Invalid parameter provided. Cannot apply to ${parameters.appliedTo}`
      );
    }
    return super.play(parameters);
  }

  toJSON(): Pojo {
    return {
      ...super.toJSON(),
      appliesTo: this.appliesTo,
    };
  }
}

export class BreakActionCard extends ToolActionCard {}

export class RepairActionCard extends ToolActionCard {}
