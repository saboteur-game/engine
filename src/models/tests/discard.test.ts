import Discard from "../discard";
import { MapActionCard } from "../cards/action-cards";

describe("Discard", () => {
  let card: MapActionCard;
  let discard: Discard;

  beforeEach(() => {
    card = new MapActionCard();
    discard = new Discard();
  });

  it("can be serialized", () => {
    expect(discard.toJS()).toMatchSnapshot();
  });

  describe("adding a played card", () => {
    it("adds the card to the discard pile", () => {
      expect(discard.getCardCount()).toBe(0);
      expect(card.status).toBe("unused");

      discard.addPlayed(card);

      expect(discard.getCardCount()).toBe(1);
      expect(card.status).toBe("played");
    });
  });

  describe("adding a discarding a card", () => {
    it("adds the card to the discard pile", () => {
      expect(discard.getCardCount()).toBe(0);
      expect(card.status).toBe("unused");

      discard.addDiscarded(card);

      expect(discard.getCardCount()).toBe(1);
      expect(card.status).toBe("discarded");
    });
  });
});
