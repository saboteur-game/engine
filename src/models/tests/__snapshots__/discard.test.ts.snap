// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discard can be serialized 1`] = `
Object {
  "discardPile": Array [],
  "id": "test-id",
}
`;
